### PR TITLE
Request interfaces refactor

### PR DESCRIPTION
**Pull request by jimmyjames**
### Changes

Refactors the `com.auth0.android.request` interfaces, with the following goals:

- Enable custom headers or request parameters to be added to any request
- Remove redundant interfaces

Specific changes in be found in the migration guide diff. In summary:

- Methods from `ParameterizedRequest` to add headers and request parameters have been moved to the top-level `Request` interface.
- Any implementations of `Request` that required implementation of headers/params have been added.
- The `ParameterizedRequest` and `AuthRequest` interfaces have been removed, as they are no longer needed (they were added to provide some additional request types the ability to add headers/params)
- Any methods in `AuthenticationAPIClient` that returned either `AuthRequest` or `ParameterizedRequest` have been changed to return `AuthenticationRequest` or `Request`, respectively.
- Constructors or methods that accepted either of the removed types have been changed to the supported types.
- Tests updated to reflect the removal of the interfaces and the new method return types, as well as new tests added for request implementation classes that now implement the header/params methods.
- The migration guide has been updated to reflect these changes.

### Testing

Please describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. Since this library has unit testing, tests should be added for new functionality and existing tests should complete without errors.

- [ ] This change adds unit test coverage

- [ ] This change adds integration test coverage

- [ ] This change has been tested on the latest version of the platform/language or why not

### Checklist

- [X] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)

- [X] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)

- [X] All existing and new tests complete without errors

### CODE REVIEW

File: auth0/src/main/java/com/auth0/android/request/Request.java [hunk 1]
```diff
@@ -52,4 +54,33 @@
...
     T execute() throws Auth0Exception;
+
+    /**
+     * Add parameters to the request as a Map of Object with the keys as String
+     *
+     * @param parameters to send with the request
+     * @return itself
+     */
+    @NonNull
+    Request<T, U> addParameters(@NonNull Map<String, Object> parameters);
```
Comment (lbalmaceda) [inline auth0/src/main/java/com/auth0/android/request/Request.java:65]: how useful is keeping this method vs iteratively calling `addParameter(name, value)`? Maybe we should consider removing it today and adding it back later if we find a use case.

Comment (jimmyjames) [reply auth0/src/main/java/com/auth0/android/request/Request.java:65]: Discussed - given that it's already present, and does have some convenience, we're going to keep it.

File: V2_MIGRATION_GUIDE.md [hunk 2]
```diff
@@ -6,32 +6,132 @@
...
 
 - The `com.auth0.android.util.Base64` class has been removed. Use `android.util.Base64` instead.
+- The `com.auth0.android.request.ParameterizableRequest` interface has been removed. The ability to add request headers and parameters has been moved to the `com.auth0.android.request.Request` interface.
+- The `com.auth0.android.request.AuthRequest` interface has been removed. The `com.auth0.android.request.AuthenticationRequest` interface can be used instead.
 
-## Constructors removed
+### Class changes
 
-- `public SignUpRequest(@NonNull DatabaseConnectionRequest<DatabaseUser, AuthenticationException> signUpRequest, @NonNull AuthenticationRequest authenticationRequest)`. Use `public SignUpRequest(@NonNull DatabaseConnectionRequest<DatabaseUser, AuthenticationException> signUpRequest, @NonNull AuthRequest authRequest)` instead.
+- `SignupRequest` now implements `AuthencticationRequest` instead of the now-removed `AuthRequest`
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply V2_MIGRATION_GUIDE.md]: Pushed a new commit to fix

File: V2_MIGRATION_GUIDE.md [hunk 3]
```diff
@@ -6,32 +6,132 @@
...
+- `SimpleRequest` now implements `Request` instead of the now-removed `ParameterizableRequest`
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply V2_MIGRATION_GUIDE.md]: Pushed a new commit to fix

File: auth0/src/main/java/com/auth0/android/authentication/request/DelegationRequest.java [hunk 4]
```diff
@@ -69,6 +68,13 @@ public DelegationRequest(@NonNull ParameterizableRequest<T, AuthenticationExcept
         return this;
     }
 
+    @NonNull
+    @Override
+    public DelegationRequest<T> addParameter(@NonNull String name, @NonNull Object value) {
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply auth0/src/main/java/com/auth0/android/authentication/request/DelegationRequest.java:73]: Discussed - it's something we can consider going forward. For this PR, keeping as interfaces to keep the refactor manageable.

**Review by lbalmaceda — APPROVED** [review]

**Pull request merged**